Re-enable CoreLib analyzers

ViktorHofer opened the pull request:
CoreLib analyzers were unintentionally disabled with
6afe96cc2e950379d13a87efa1fa5a651cbce9bd. This re-enables the analyzers
and conditions one which brings in external dependencies that CoreLib
can't deal with.

Fixes https://github.com/dotnet/runtime/issues/422

ViktorHofer committed:
Re-enable CoreLib analyzers

CoreLib analyzers were unintentionally disabled with
6afe96cc2e950379d13a87efa1fa5a651cbce9bd. This re-enables the analyzers
and conditions one which brings in external dependencies that CoreLib
can't deal with.

stephentoub commented on eng/Analyzers.props:
```diff
@@ -3,9 +3,10 @@
-    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.6" />
-    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.66" />
+    <PackageReference Include="Microsoft.DotNet.CodeAnalysis" Version="$(MicrosoftDotNetCodeAnalysisVersion)" PrivateAssets="all" IsImplicitlyDefined="true" />
+    <!-- Microsoft.CodeAnalysis.Common brings in external dependencies that CoreLib can't deal with. -->
```
How does "can't deal with" manifest?

ViktorHofer replied:
Thousands of the following errors:

```
C:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\MemoryFailPoint.cs(368,128): error CS0518: Predefined type 'System.Boolean' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
C:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\MemoryFailPoint.cs(368,156): error CS0518: Predefined type 'System.UInt64' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
C:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\MemoryFailPoint.cs(368,177): error CS0518: Predefined type 'System.UInt64' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
C:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\MemoryFailPoint.cs(368,206): error CS0518: Predefined type 'System.Int64' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
C:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\MemoryFailPoint.cs(368,238): error CS0518: Predefined type 'System.UInt64' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
C:\git\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\MemoryFailPoint.cs(368,13): error CS0518: Predefined type 'System.Void' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
src\System\RuntimeArgumentHandle.cs(15,23): error CS0518: Predefined type 'System.ValueType' is not defined or imported [C:\git\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
```

AFAIK CoreLib mustn't have any references. That packages brings some in. I'm right now experimenting with adding a netstandard2.1 configuration to `Microsoft.CodeAnalysis.csproj`.

jkotas replied:
The analyzer dependencies are not supposed to end up as `/r:` references for the compiler. This means that either this package is not authored correctly; or it is not used correctly.

ViktorHofer replied:
> or it is not used correctly.

It shouldn't be referenced directly. The entry was added because libraries inject the analyzers into the csc task in a special way (no project restore). This should now be resolved.

jkotas commented on eng/Analyzers.props:
```diff
@@ -3,9 +3,10 @@
-    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.6" />
-    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.66" />
+    <PackageReference Include="Microsoft.DotNet.CodeAnalysis" Version="$(MicrosoftDotNetCodeAnalysisVersion)" PrivateAssets="all" IsImplicitlyDefined="true" />
+    <!-- Microsoft.CodeAnalysis.Common brings in external dependencies that CoreLib can't deal with. -->
```
What is the `Microsoft.CodeAnalysis.Common` package for? What is not going to work if we remove from here?

ViktorHofer replied:
That package was first introduced with https://github.com/dotnet/corefx/commit/87873f55d4d0d4f2b073e544f1295761976beca8#diff-c9ed7fc6ec49880569e6c2eb08120fc4R27. I'm not exactly sure, its description even says it shouldn't be referenced directly. I'll take a closer look.

ViktorHofer commented:
I'll remove that package as its description clearly state that it shouldn't be referenced directly. @safern do you have context why that package was referenced directly in corefx/arcade?

ViktorHofer committed:
Remove Microsoft.CodeAnalysis.Common from common

Microsoft.CodeAnalysis.Common shouldn't be referenced directly,
therefore removing that package from the common location and only
restoring and marking it as an analyzer in the libraries restore phase.

ViktorHofer commented:
OK, so the transitive dependency to Microsoft.CodeAnalysis.Common is required becaues of the special way we add analyzers to projects in libraries. We will clean this up with @Anipik's work to enable project restore for non test projects in libraries.

jkotas commented on src/libraries/restore/analyzers/analyzers.depproj:
```diff
@@ -5,6 +5,14 @@
+  <ItemGroup>
+    <!-- 
+      Microsoft.CodeAnalysis.Common shouldn't be referenced directly but is needed by the analyzer depproj restore.
+      TODO: Remove when project restore is enabled for non test projects.
```
Add link to the issue that tracks this work?

ViktorHofer replied:
done.

jkotas commented on eng/Analyzers.props:
```diff
@@ -3,9 +3,8 @@
-    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="3.4.0-beta2-final" />
-    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.6" />
-    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.66" />
+    <PackageReference Include="Microsoft.DotNet.CodeAnalysis" Version="$(MicrosoftDotNetCodeAnalysisVersion)" PrivateAssets="all" IsImplicitlyDefined="true" />
```
Why this one needs `IsImplicitlyDefined="true"`, but the other two don't?

ViktorHofer replied:
Because that package is part of the dependency flow and shouldn't be updatable via the nuget package manager api. The other two are not part of the dependency flow.

ViktorHofer committed:
Add todo link

ViktorHofer commented:
This should be ready, can someone pls approve? Thanks

ViktorHofer commented:
Thanks, merging as the previous run was green (except an unrelated test failure) and I just changed a comment which re-triggered CI.

safern reviewed (commented):
LGTM
